include/aws/crt/mqtt/Mqtt5Listener.h

```diff
@@ -0,0 +1,344 @@
+
+                static bool s_publishReceivedCallback(const aws_mqtt5_packet_publish_view *publish, void *user_data);
+
+                static void s_listenerTerminationCompletion(void *complete_ctx);
+
+                /**
+                 * Callback handler triggered when client successfully establishes an MQTT connection
+                 */
+                OnConnectionSuccessHandler onConnectionSuccess;
+
+                /**
+                 * Callback handler triggered when client fails to establish an MQTT connection
+                 */
+                OnConnectionFailureHandler onConnectionFailure;
+
+                /**
+                 * Callback handler triggered when client's current MQTT connection is closed
+                 */
+                OnDisconnectionHandler onDisconnection;
+
+                /**
+                 * Callback handler triggered when client reaches the "Stopped" state
+                 */
+                OnStoppedHandler onStopped;
+
+                /**
+                 * Callback handler triggered when client begins an attempt to connect to the remote endpoint.
+                 */
+                OnAttemptingConnectHandler onAttemptingConnect;
+
+                /**
+                 * Callback handler triggered when an MQTT PUBLISH packet is received by the client
+                 */
+                OnListenerPublishReceivedHandler onListenerPublishReceived;
+
+                /*
+                 * Reference to Mqtt5Client. As the listener has a dependency on mqtt5_client, we would like to keep the
+                 * Mqtt5Client alive.
+                 */
+                std::shared_ptr<Mqtt5Client> m_mqtt5Client;
+
+                /*
+                 * Self reference to keep self alive until the termination callback is invoked.
+                 */
+                std::shared_ptr<Mqtt5ListenerCore> m_selfReference;
+
+                /*
+                 * The callback flag used to indicate if it is safe to invoke the callbacks
+                 */
+                enum CallbackFlag
+                {
+                    INVOKE,
+                    IGNORE
+                } m_callbackFlag;
+
+                /*
+                 * Lock for the callbacks. This is used to protect the callback flag and callbacks.
+                 */
+                std::recursive_mutex m_callback_lock;
+
+                /**
+                 * Underlying c classes for data storage and operational members
+                 */
+                aws_mqtt5_listener *m_listener;
+
+                std::map<Crt::String, Crt::Mqtt5::OnPublishReceivedHandler> m_subscriptionMap;
+
+                Allocator *m_allocator;
+            };
+
+            /**
+             * An MQTT5 listener.
+             *
+             * This is a move-only type. Unless otherwise specified, all function arguments need only to live through
```

Comment:
trivial
"all function arguments only need to live through the duration of the function call"